src/hardware/pcspeaker.cpp

```diff
@@ -274,34 +277,28 @@ void PCSPEAKER_SetType(Bitu mode)
 	spkr.chan->Enable(true);
 }
 
-// Once the speaker is idle (as defined when num_actions is empty),
-// give it a short grace-time before ramping the volume down.
-// By keeping this idle-phase relatively short, we reduce the chance
-// PC-Speaker's DC-offset can influence other channels (such as the
-// Sound blaster or GUS).  Counts are in milliseconds.
-static void FadeVolume(uint16_t num_actions)
+// After the speaker has idled (as defined when the number of speaker
+// movements is zero) for a short time, wind down the DC-offset, and
+// then halt the channel.
+static void PlayOrSilence(const uint16_t speaker_movements,
+                          size_t requested_samples,
+                          int16_t *buffer)
 {
-	constexpr uint8_t grace_start = 50u;
-	constexpr uint8_t grace_end = 15u;
-	assert(grace_start >= grace_end);
-	static_assert(grace_end > 0, "algorithm depends on non-zero grace_end value");
+	constexpr uint8_t grace_time_ms = 100u;
+	static_assert(grace_time_ms > 0,
+	              "Algorithm depends on a non-zero grace time");
 
-	// If the speaker was used then reset our countdown.
-	if (num_actions) {
-		spkr.idle_countdown = grace_start;
-		return;
-	}
-	// If the count is lower than the idle period, then ramp down the volume.
-	if (spkr.idle_countdown && spkr.idle_countdown < grace_end) {
-		spkr.volwant = spkr.volwant * spkr.idle_countdown / grace_end;
+	if (speaker_movements && requested_samples) {
+		spkr.idle_countdown = grace_time_ms;
+		spkr.dc_silencer.Reset();
+	} else if (spkr.idle_countdown) {
 		spkr.idle_countdown--;
-	}
-	// If we've ramped to zero, then shutdown the channel. This function
-	// will not be called again until the channel's re-enabled through
-	// the PWM or PIT functions.
-	if (!spkr.idle_countdown && SpeakerExists()) {
+		spkr.last_played_sample = buffer[requested_samples - 1];
+	} else if (!spkr.dc_silencer.Generate(spkr.last_played_sample,
+	                                    requested_samples, buffer))
```

Comment:
Please add braces, because this is a chain "if / else if / else if" (with all earlier blocks using braces).

Also, 2 missing spaces in line 298 ;)

Reply:
good eye :-) Got 'em.